File: frontend/src/lib/socReadiness.js
import {recordUuid} from './recordUuid';
export const SOC_CATEGORIES={security:'Security / Common Criteria',availability:'Availability',confidentiality:'Confidentiality',processing_integrity:'Processing Integrity',privacy:'Privacy'};
export const socConfiguration=client=>({categories:['security'],system_description:'',period_start:'',period_end:'',...client?.framework_settings?.['soc-2']});
export const controlGap=control=>control.expected_instances==null||control.collected_instances==null?null:Math.max(0,control.expected_instances-control.collected_instances);
export const newManagementControl=(period={})=>({control_id:recordUuid(),name:'',description:'',design:'not_assessed',operating:'not_assessed',frequency:'',period_start:period.period_start||'',period_end:period.period_end||'',expected_instances:null,collected_instances:null,population_notes:'',testing_notes:''});

function period(start,end){
  if(!start&&!end)return;
  if(!start||!end)throw new Error('Provide both evidence-period dates');
  for(const value of [start,end]){
    const parsed=new Date(value+'T00:00:00Z');
    if(!/^\d{4}-\d{2}-\d{2}$/.test(value)||Number.isNaN(parsed.getTime())||parsed.toISOString().slice(0,10)!==value)throw new Error('Use valid YYYY-MM-DD evidence-period dates');
  }
  if(end<start)throw new Error('Evidence period must end on or after its start');
}
function text(value,max,label){
  if(typeof value!=='string'||value.trim().length>max)throw new Error('Invalid '+label);
  return value.trim();
}
export function validateSocConfiguration(body){
  if(Object.keys(body).some(k=>!['client_id','categories','system_description','period_start','period_end'].includes(k)))throw new Error('Unknown SOC 2 configuration fields');
  const cid=text(body.client_id,160,'client');
  const categories=body.categories===undefined?['security']:body.categories;
  if(!cid||!Array.isArray(categories)||categories.length>5||!categories.includes('security')||new Set(categories).size!==categories.length||categories.some(c=>!SOC_CATEGORIES[c]))throw new Error('Common Criteria must remain in scope; categories must be valid and unique');
  const config={categories,system_description:text(body.system_description===undefined?'':body.system_description,4000,'system description'),period_start:text(body.period_start===undefined?'':body.period_start,10,'period start'),period_end:text(body.period_end===undefined?'':body.period_end,10,'period end')};
  period(config.period_start,config.period_end);return config;
}
export function validateManagementControls(values){
  if(!Array.isArray(values)||values.length>30)throw new Error('Use at most 30 management control descriptions per criterion');
  const fields=['control_id','name','description','design','operating','frequency','period_start','period_end','expected_instances','collected_instances','population_notes','testing_notes'];
  const normalized=values.map(c=>{
    if(!c||typeof c!=='object'||Object.keys(c).some(k=>!fields.includes(k)))throw new Error('Unknown management control fields');
    const result={control_id:text(c.control_id,80,'control identifier'),name:text(c.name,200,'control name'),
      design:c.design===undefined?'not_assessed':c.design,operating:c.operating===undefined?'not_assessed':c.operating};
    if(!result.control_id||!result.name||!['not_assessed','adequate','gap'].includes(result.design)||!['not_assessed','effective','gap'].includes(result.operating))throw new Error('Provide a control name and valid readiness states');
    for(const k of ['description','frequency','period_start','period_end','population_notes','testing_notes'])result[k]=text(c[k]===undefined?'':c[k],k.startsWith('period_')?10:k==='frequency'?200:4000,k);
    for(const k of ['expected_instances','collected_instances']){
      const value=c[k]??null;
      if(value!==null&&(!Number.isInteger(value)||value<0||value>1000000))throw new Error('Instance counts must be whole numbers from 0 to 1000000');
      result[k]=value;
    }
    period(result.period_start,result.period_end);
    if((result.expected_instances!==null||result.collected_instances!==null)&&!result.period_start)throw new Error('Instance counts require a defined evidence period');
    return result;
  });
  if(new Set(normalized.map(c=>c.control_id)).size!==normalized.length)throw new Error('Management control identifiers must be unique');
  return normalized;
}

File: backend/soc_readiness.py
"""Bounded internal readiness input, not audit sampling or attestation logic."""
from datetime import date
import re
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, Field, model_validator

Category = Literal['security','availability','confidentiality','processing_integrity','privacy']


def validate_period(start, end):
    if not start and not end:
        return
    if not start or not end:
        raise ValueError('Provide both evidence-period dates')
    for value in (start, end):
        if not re.fullmatch(r'\d{4}-\d{2}-\d{2}', value):
            raise ValueError('Use YYYY-MM-DD evidence-period dates')
        date.fromisoformat(value)
    if end < start:
        raise ValueError('Evidence period must end on or after its start')


class SocConfiguration(BaseModel):
    model_config = ConfigDict(extra='forbid', str_strip_whitespace=True)
    client_id: str = Field(min_length=1, max_length=160)
    expected_updated_at: Optional[str] = Field(default=None,max_length=100)
    categories: list[Category] = Field(default_factory=lambda: ['security'], min_length=1, max_length=5)
    system_description: str = Field(default='', max_length=4000)
    period_start: str = Field(default='', max_length=10)
    period_end: str = Field(default='', max_length=10)

    @model_validator(mode='after')
    def consistent_scope(self):
        if 'security' not in self.categories or len(set(self.categories)) != len(self.categories):
            raise ValueError('Common Criteria must remain in scope; categories must be unique')
        validate_period(self.period_start, self.period_end)
        return self


class ManagementControl(BaseModel):
    model_config = ConfigDict(extra='forbid', str_strip_whitespace=True)
    control_id: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default='', max_length=4000)
    design: Literal['not_assessed','adequate','gap'] = 'not_assessed'
    operating: Literal['not_assessed','effective','gap'] = 'not_assessed'
    frequency: str = Field(default='', max_length=200)
    period_start: str = Field(default='', max_length=10)
    period_end: str = Field(default='', max_length=10)
    expected_instances: Optional[int] = Field(default=None, strict=True, ge=0, le=1000000)
    collected_instances: Optional[int] = Field(default=None, strict=True, ge=0, le=1000000)
    population_notes: str = Field(default='', max_length=4000)
    testing_notes: str = Field(default='', max_length=4000)

    @model_validator(mode='after')
    def consistent_period(self):
        validate_period(self.period_start, self.period_end)
        if (self.expected_instances is not None or self.collected_instances is not None) and not self.period_start:
            raise ValueError('Instance counts require a defined evidence period')
        return self


def configuration(client):
    return {'categories':['security'], 'system_description':'', 'period_start':'', 'period_end':'',
            **client.get('framework_settings', {}).get('soc-2', {}),
            'expected_updated_at':client.get('soc_configuration_updated_at')}
